libc/stdlib_env.c

```diff
@@ -0,0 +1,295 @@
+			else { /* var not set must be null */
+				TEST_ASSERT_NULL(var);
+			}
+		}
+	}
+}
+
+
+TEST(stdlib_env, empty_name)
+{
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, setenv("", TEST_VALUE, 0));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, setenv("", TEST_VALUE, 1));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+	char *var = getenv("");
+	TEST_ASSERT_NULL(var);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, unsetenv(""));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+}
+
+
+TEST(stdlib_env, empty_null)
+{
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, setenv(TEST_VAR, "", 0));
+	TEST_ASSERT_NOT_EQUAL_INT(EINVAL, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, setenv(TEST_VAR, "", 1));
+	TEST_ASSERT_NOT_EQUAL_INT(EINVAL, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, setenv(NULL, TEST_VALUE, 0));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+}
+
+
+TEST(stdlib_env, putenv_invalid)
+{
+#ifndef __phoenix__
+	TEST_ASSERT_EQUAL_INT(0, putenv(""));
+#endif
+
+#ifdef __phoenix__
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, putenv(""));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+	TEST_ASSERT_EQUAL_INT(-1, putenv("abc"));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+#endif
+}
+
+
+TEST(stdlib_env, invalid_value)
+{
+	int i;
+	for (i = 0; i < 2; i++) {
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, setenv(INVALID_VAR1, INVALID_VAL1, i));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, setenv(INVALID_VAR2, INVALID_VAL2, i));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, setenv(INVALID_VAR3, INVALID_VAL3, i));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, unsetenv(INVALID_VAR1));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, unsetenv(INVALID_VAR2));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(-1, unsetenv(INVALID_VAR3));
+		TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+	}
+}
+
+
+TEST(stdlib_env, test_overwrite)
+{
+	TEST_ASSERT_EQUAL_INT(0, setenv(TEST_VAR, TEST_VALUE, 1));
+
+	TEST_ASSERT_EQUAL_INT(0, setenv(TEST_VAR, TEST_VALUE2, 1));
+
+	TEST_ASSERT_EQUAL_STRING(TEST_VALUE2, getenv(TEST_VAR));
+}
+
+
+TEST(stdlib_env, env_not_set)
+{
+	TEST_ASSERT_NULL(getenv(NOT_SET_VAR));
+}
+
+TEST(stdlib_env, stdlib_environ_len)
+{
+	int len, len2;
+	len = get_environ_len(environ);
+	TEST_ASSERT_EQUAL_INT(0, setenv("len_test", TEST_VALUE, 1));
+	len2 = get_environ_len(environ);
+	TEST_ASSERT_EQUAL_INT(len + 1, len2);
+
+	TEST_ASSERT_EQUAL_INT(0, unsetenv("len_test"));
+	len2 = get_environ_len(environ);
+	TEST_ASSERT_EQUAL_INT(len, len2);
+}
+
+
+TEST_GROUP_RUNNER(stdlib_env)
+{
+	RUN_TEST_CASE(stdlib_env, random);
+	RUN_TEST_CASE(stdlib_env, invalid_value);
+	RUN_TEST_CASE(stdlib_env, empty_name);
+	RUN_TEST_CASE(stdlib_env, empty_null);
+	RUN_TEST_CASE(stdlib_env, putenv_invalid);
+	RUN_TEST_CASE(stdlib_env, env_not_set);
+	RUN_TEST_CASE(stdlib_env, stdlib_environ_len);
```

Comment:
In case of error it will be much easier to locate the cause of error in smaller cases. That's why we can call `random` test case at the end of a test. Above that you can add `simple` test case, which will be called first, with a few simple examples of setting/getting/putting and unsetting some env variables. Then you can create and call `long` test case, which will test some longer names/values, next maybe `ascii` with all ascii characters possible to use (both in name and value). After that you can call the rest of your current cases and `random` at the end.

```suggestion
	RUN_TEST_CASE(stdlib_env, invalid_value);
	RUN_TEST_CASE(stdlib_env, empty_name);
	RUN_TEST_CASE(stdlib_env, empty_null);
	RUN_TEST_CASE(stdlib_env, putenv_invalid);
	RUN_TEST_CASE(stdlib_env, env_not_set);
	RUN_TEST_CASE(stdlib_env, stdlib_environ_len);
	RUN_TEST_CASE(stdlib_env, random);
```